src/main/scala/org/clulab/habitus/utils/ContextExtractor.scala

```diff
@@ -0,0 +1,260 @@
+package org.clulab.habitus.utils
+
+import org.clulab.habitus.variables.EntityDistFreq
+
+import org.clulab.odin.{Attachment, Mention}
+import org.clulab.processors.Document
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+trait Context extends Attachment {
+  def getArgValuePairs() = this.getClass.getDeclaredFields.toList
+    .map(arg => {
+      arg.setAccessible(true)
+      (arg.getName, arg.get(this))
+    })
+
+  def getTSVContextHeader() = {
+    getArgValuePairs().map(_._1).mkString("\t")
+  }
+
+  def getTSVContextString() = {
+    getArgValuePairs().map(_._2).mkString("\t")
+  }
+}
+
+case class DefaultContext(location: String, date: String, crop: String, fertilizer: String, comparative: Int) extends Context
+
+trait ContextExtractor {
+
+  def getContextPerMention(mentions: Seq[Mention], entityHistogram: Seq[EntityDistFreq], doc: Document, label: String): Seq[Mention]
+
+  def getComparative(mention: Mention): Int = {
+    val relative = Seq("vs", "vs.", "respectively")
+    if (mention.sentenceObj.words.intersect(relative).nonEmpty) 1 else 0
+  }
+
+  def getLocation(m: Mention, thisSentLocs: Seq[Mention], frequencyContext: Map[Int, ContextDetails]): String = {
+    val location = thisSentLocs.length match {
+      // if no locations in sentence, use the most freq one in +/- 1 sent window
+      case 0 => if (frequencyContext.nonEmpty) {
+        frequencyContext(m.sentence).mostFreqLoc1Sent
+      } else "N/A"
+      case 1 => thisSentLocs.head.text
+      case _ => {
+        val nextLoc = findClosestNextLocation(m, thisSentLocs)
+        if (nextLoc != null) nextLoc.text else "N/A"
+      }
+    }
+    location
+  }
+
+  def getCropContext(m: Mention, frequencyContext: Map[Int, ContextDetails]): String = {
+    if (frequencyContext.nonEmpty) {
+      frequencyContext(m.sentence).mostFreqCrop
+    } else "N/A"
+  }
+
+  def getFertilizerContext(m: Mention, frequencyContext: Map[Int, ContextDetails]): String = {
+    if (frequencyContext.nonEmpty) {
+      frequencyContext(m.sentence).mostFreqFertilizerOverall
+    } else "N/A"
+  }
+
+  def getDate(m: Mention, thisSentDates: Seq[Mention], frequencyContext: Map[Int, ContextDetails]): String = {
+    // if no dates in sentence, use the most freq one in +/- 1 sent window
+    val date = thisSentDates.length match {
+      case 0 => if (frequencyContext.nonEmpty) {
+        frequencyContext(m.sentence).mostFreqDate1Sent
+      } else "N/A"// because there are no dates in this sentence, take most freq in -/+ 1 window
```

Comment:
I'd be likely to mistype one of these or decide to change it one day and miss one, therefore I'm forced to write
```scala
val NA = "N/A"
```

Reply:
Yes!